### PR TITLE
[cli] Fix 'vc remove --safe <project>'

### LINKED ISSUE #9321: `vercel rm <project> --safe` is not working
The `vercel rm <project> --safe` command is not working as intended in the latest version of the CLI (currently v28.13.1).

### Steps to reproduce:

1. Create a few deployments
2. Run the command `vercel rm <project> --safe`
3. Receive the incorrect error:
    ```    
    > Could not find unaliased deployments or projects matching "<project>". Run `vercel ls` to list.
    ```

### Expected:

The command should print all deployments except for the current production deployment and prompt you to delete.  Example:
```
> Found N deployments for removal in <scope> [723ms]
> The following N deployments will be permanently removed:
  dpl_<id>      <url>     11h ago
  ...
> Are you sure? [y/N]
```

To confirm this expected behavior, I ran the command with a version of the CLI from several months ago (v28.4.8) and the command worked as expected.

### Additional Information

Here is the debug output from each CLI version when running these commands:

<details>
<summary>Working (v28.4.8)</summary>

```
> [debug] [2023-01-26T05:30:45.414Z] user supplied known subcommand: "rm"
> [debug] [2023-01-26T05:30:45.434Z] #1 → GET https://api.vercel.com/v2/user
> [debug] [2023-01-26T05:30:45.590Z] #1 ← 200 OK: sfo1::mxm5k-1674711045582-e0f2f1d06c51 [156ms]
> [debug] [2023-01-26T05:30:45.592Z] Spinner invoked (Fetching deployment(s) "<project>" in <user>) with a 300ms delay
> [debug] [2023-01-26T05:30:45.596Z] #2 → GET https://api.vercel.com/v5/now/deployments/<project>?
> [debug] [2023-01-26T05:30:45.598Z] #3 → GET https://api.vercel.com/v8/projects/<project>?
> [debug] [2023-01-26T05:30:45.747Z] #2 ← 404 Not Found: sfo1::rt842-1674711045747-041940a27823 [151ms]
> [debug] [2023-01-26T05:30:45.765Z] #3 ← 200 OK: sfo1::pjfmt-1674711045743-da54e38ad74e [167ms]
> [debug] [2023-01-26T05:30:45.771Z] #4 → GET https://api.vercel.com/v4/now/deployments?projectId=prj_<id>&limit=100
> [debug] [2023-01-26T05:30:45.935Z] #4 ← 200 OK: sfo1::25qb8-1674711045913-b68a0f7843e5 [163ms]
> [debug] [2023-01-26T05:30:45.943Z] #5 → GET https://api.vercel.com/now/deployments/dpl_<id>/aliases?
> [debug] [2023-01-26T05:30:45.947Z] #6 → GET https://api.vercel.com/now/deployments/dpl_<id>/aliases?
> [debug] [2023-01-26T05:30:45.950Z] #7 → GET https://api.vercel.com/now/deployments/dpl_<id>/aliases?
> [debug] [2023-01-26T05:30:46.172Z] #5 ← 200 OK: sfo1::5rpvc-1674711046085-392d8e79c6a5 [229ms]
> [debug] [2023-01-26T05:30:46.180Z] #7 ← 200 OK: sfo1::n8tp4-1674711046096-1290a6aa6f33 [230ms]
> [debug] [2023-01-26T05:30:46.183Z] #6 ← 200 OK: sfo1::qncbd-1674711046096-e79353408dfa [236ms]
```
</details>

<details>
<summary>Not working (v28.13.1 latest)</summary>

```
> [debug] [2023-01-26T05:30:20.722Z] user supplied known subcommand: "rm"
> [debug] [2023-01-26T05:30:20.744Z] #1 → GET https://api.vercel.com/v2/user
> [debug] [2023-01-26T05:30:21.032Z] #1 ← 200 OK: sfo1::8xsng-1674711021021-f4a5ccad3cbc [288ms]
> [debug] [2023-01-26T05:30:21.035Z] Spinner invoked (Fetching deployment(s) "<project>" in <user>) with a 300ms delay
> [debug] [2023-01-26T05:30:21.041Z] #2 → GET https://api.vercel.com/v13/deployments/<project>
> [debug] [2023-01-26T05:30:21.045Z] #3 → GET https://api.vercel.com/v8/projects/<project>
> [debug] [2023-01-26T05:30:21.214Z] #2 ← 404 Not Found: sfo1::pcrkt-1674711021205-b85db600fed1 [173ms]
> [debug] [2023-01-26T05:30:21.228Z] #3 ← 200 OK: sfo1::b54z9-1674711021205-8b4f8c3faed3 [183ms]
> [debug] [2023-01-26T05:30:21.235Z] #4 → GET https://api.vercel.com/v4/now/deployments?projectId=prj_<id>&limit=100
> [debug] [2023-01-26T05:30:21.423Z] #4 ← 200 OK: sfo1::b54z9-1674711021376-b4b5efb94e4c [187ms]
> [debug] [2023-01-26T05:30:21.431Z] #5 → GET https://api.vercel.com/v3/now/aliases?limit=20
> [debug] [2023-01-26T05:30:21.435Z] #6 → GET https://api.vercel.com/v3/now/aliases?limit=20
> [debug] [2023-01-26T05:30:21.437Z] #7 → GET https://api.vercel.com/v3/now/aliases?limit=20
> [debug] [2023-01-26T05:30:21.642Z] #7 ← 200 OK: sfo1::dhsr6-1674711021588-a5ed20901dbd [205ms]
> [debug] [2023-01-26T05:30:21.661Z] #5 ← 200 OK: sfo1::h58kr-1674711021578-57a823bcb98c [230ms]
> [debug] [2023-01-26T05:30:21.668Z] #6 ← 200 OK: sfo1::kslfr-1674711021593-a41f58517c08 [233ms]
```
</details>

### PR DESCRIPTION
**Pull request by cb1kenobi**
`vc remove --safe <project>` has a unique code path that queries all deployments by project id. This function calls v4 deployments endpoint and returns the legacy deployment structure.

The problem is the `vc remove` command relies on the new `Deployment` structure defined by the v13 get deployment endpoint.

I've updated the `getDeploymentsByProjectId()` function to transform the legacy structure into the current one.

Fixes #9321 

Linear: https://linear.app/vercel/issue/VCCLI-538/vc-rm-project-safe-is-not-working

### CODE REVIEW

File: packages/cli/src/util/deploy/get-deployments-by-project-id.ts [hunk 1]
```diff
@@ -45,17 +66,48 @@ export default async function getDeploymentsByProjectId(
     query.set('from', options.from.toString());
   }
 
-  const { deployments } = await client.fetch<Response>(
-    `/v4/now/deployments?${query}`
+  const { deployments: legacyDeployments } = await client.fetch<Response>(
+    `/v6/deployments?${query}`
   );
+
+  // we need to transform the old deployment shape to the new shape
```
Comment (styfle) [inline packages/cli/src/util/deploy/get-deployments-by-project-id.ts:73]: Can we instead use a newer API version to get the new Deployment type for all cases?

File: packages/cli/src/commands/remove.ts [hunk 2]
```diff
@@ -167,9 +169,15 @@ export default async function main(client: Client) {
...
       );
 
-      projectDeployments
-        .slice(0, 201)
-        .map(pDeployments => deployments.push(...pDeployments));
+      const to = Math.min(projectDeployments.length, 201);
+      for (let i = 0; i < to; i++) {
+        for (const pDepl of projectDeployments[i]) {
+          deployments.push({
+            ...pDepl,
```
Comment (styfle) [inline packages/cli/src/commands/remove.ts]: Can we avoid spread here and assign the prop instead?

File: packages/cli/src/util/deploy/get-deployments-by-project-id.ts (hunk 1 above)
Comment (cb1kenobi) [reply packages/cli/src/util/deploy/get-deployments-by-project-id.ts:73]: This is the latest version: https://vercel.com/docs/rest-api#endpoints/deployments/list-deployments.

File: packages/cli/src/commands/remove.ts (hunk 2 above)
Comment (cb1kenobi) [reply packages/cli/src/commands/remove.ts]: Fixed

File: packages/cli/src/commands/remove.ts [hunk 3]
```diff
@@ -167,9 +169,15 @@ export default async function main(client: Client) {
         })
       );
 
-      projectDeployments
-        .slice(0, 201)
-        .map(pDeployments => deployments.push(...pDeployments));
+      // only process the first 201 projects
```
Comment (EndangeredMassa) [inline packages/cli/src/commands/remove.ts:172]: **question (non-blocking):** why this limit?

This comment says what is happening, but I'm more interested in why. Is this a limit of the system somewhere?

File: packages/cli/src/commands/remove.ts [hunk 4]
```diff
@@ -167,9 +169,15 @@ export default async function main(client: Client) {
...
+      const to = Math.min(projectDeployments.length, 201);
+      for (let i = 0; i < to; i++) {
+        for (const pDepl of projectDeployments[i]) {
+          const depl = pDepl as DeploymentWithAliases;
```
Comment (EndangeredMassa) [inline packages/cli/src/commands/remove.ts:176]: **suggestion (non-blocking):** remove type assertion

Can we make this type check without a type assertion?

File: packages/cli/src/util/deploy/get-deployments-by-project-id.ts [hunk 5]
```diff
@@ -1,26 +1,47 @@
...
     | 'QUEUED'
     | 'READY'
     | 'CANCELED';
-  creator: { uid: string };
-}
+  target?: 'production' | 'staging' | null;
+  type: 'LAMBDAS';
+  uid: string;
+  url: string;
+};
```
Comment (EndangeredMassa) [inline packages/cli/src/util/deploy/get-deployments-by-project-id.ts:40]: **suggestion (non-blocking):** extract general type

This type could be useful elsewhere. Can we make it more discoverable by placing it a common place instead?

File: packages/cli/src/util/deploy/get-deployments-by-project-id.ts [hunk 6]
```diff
@@ -1,26 +1,47 @@
...
 import Client from '../client';
 
-type Response = {
-  deployments: DeploymentPartial[];
-};
-export interface DeploymentPartial {
-  id: string;
-  name: string;
-  url: string;
+type LegacyDeployment = {
```
Comment (EndangeredMassa) [inline packages/cli/src/util/deploy/get-deployments-by-project-id.ts:5]: **suggestion (non-blocking):** change type name to be more specific

What does "legacy" mean, specifically? It looks like this was an unnamed `version: 1` and the current `Deployment` is `version: 2`, right?

If that's correct, can we call this `DeploymentV1` and the modern version `DeploymentV2` to keep them clear?

File: packages/cli/src/util/deploy/get-deployments-by-project-id.ts (hunk 1 above)
Comment (EndangeredMassa) [reply packages/cli/src/util/deploy/get-deployments-by-project-id.ts:73]: So the latest version of `list-deployments` returns an old version of `Deployment` items? Can we ask Build and Deploy to look into creating a new version of the endpoint that responds with the latest version of `Deployment`?

File: packages/cli/src/util/deploy/get-deployments-by-project-id.ts [hunk 7]
```diff
@@ -45,17 +66,48 @@ export default async function getDeploymentsByProjectId(
...
+      aliasError: depl.aliasError,
+      buildingAt: depl.buildingAt,
+      checksConclusion: depl.checksConclusion,
+      checksState: depl.checksState,
+      createdAt: depl.created,
+      creator: {
+        uid: depl.creator.uid,
+        username: depl.creator.username,
+      },
+      id: depl.uid,
```
Comment (EndangeredMassa) [inline packages/cli/src/util/deploy/get-deployments-by-project-id.ts:86]: **question (non-blocking):** do we have to check for `.id` as well here?

**Review by EndangeredMassa — APPROVED** [review]

File: packages/cli/src/commands/remove.ts (hunk 3 above)
Comment (cb1kenobi) [reply packages/cli/src/commands/remove.ts:172]: I have no idea. I was curious about that limit also, but I have no idea hence the comment. I'm happy to remove the `.slice()` if there's no reason for the limit.

File: packages/cli/src/commands/remove.ts (hunk 4 above)
Comment (cb1kenobi) [reply packages/cli/src/commands/remove.ts:176]: I tried that. Actually, I believe I originally used the spread operator to copy the `Deployment` into the new `DeploymentWithAliases` type, but was an unnecessary performance hit that casting doesn't have.

**Pull request merged**